librz/util/print.c

```diff
@@ -512,56 +512,180 @@ RZ_API void rz_print_section(RzPrint *p, ut64 at) {
-	bool c = p ? (p->flags & RZ_PRINT_FLAGS_COLOR) : false;
-	const char *color_title = c ? (Pal(p, offset)
-					      : Color_MAGENTA)
-				    : "";
-	int inc = p ? p->cols : 16;
-	size_t i, j, k;
+static inline void print_addr(RzStrBuf *sb, RzPrint *p, ut64 addr) {
+	char space[32] = {
+		0
+	};
+	const char *white = "";
+#define PREOFF(x) (p && p->cons && p->cons->context && p->cons->context->pal.x) ? p->cons->context->pal.x
+	bool use_segoff = p ? (p->flags & RZ_PRINT_FLAGS_SEGOFF) : false;
+	bool use_color = p ? (p->flags & RZ_PRINT_FLAGS_COLOR) : false;
+	bool dec = p ? (p->flags & RZ_PRINT_FLAGS_ADDRDEC) : false;
+	bool mod = p ? (p->flags & RZ_PRINT_FLAGS_ADDRMOD) : false;
+	char ch = p ? ((p->addrmod && mod) ? ((addr % p->addrmod) ? ' ' : ',') : ' ') : ' ';
+	if (p && p->flags & RZ_PRINT_FLAGS_COMPACT && p->col == 1) {
+		ch = '|';
+	}
+	if (p && p->pava) {
+		ut64 va = p->iob.p2v(p->iob.io, addr);
+		if (va != UT64_MAX) {
+			addr = va;
+		}
+	}
+	if (use_segoff) {
+		ut32 s, a;
+		a = addr & 0xffff;
+		s = (addr - a) >> (p ? p->seggrn : 0);
+		if (dec) {
+			snprintf(space, sizeof(space), "%d:%d", s & 0xffff, a & 0xffff);
+			white = rz_str_pad(' ', 9 - strlen(space));
+		}
+		if (use_color) {
+			const char *pre = PREOFF(offset)
+			    : Color_GREEN;
+			const char *fin = Color_RESET;
+			if (dec) {
+				rz_strbuf_appendf(sb, "%s%s%s%s%c", pre, white, space, fin, ch);
+			} else {
+				rz_strbuf_appendf(sb, "%s%04x:%04x%s%c", pre, s & 0xffff, a & 0xffff, fin, ch);
+			}
+		} else {
+			if (dec) {
+				rz_strbuf_appendf(sb, "%s%s%c", white, space, ch);
+			} else {
+				rz_strbuf_appendf(sb, "%04x:%04x%c", s & 0xffff, a & 0xffff, ch);
+			}
+		}
+	} else {
+		if (dec) {
+			snprintf(space, sizeof(space), "%" PFMT64d, addr);
+			int w = RZ_MAX(10 - strlen(space), 0);
+			white = rz_str_pad(' ', w);
+		}
+		if (use_color) {
+			const char *pre = PREOFF(offset)
+			    : Color_GREEN;
+			const char *fin = Color_RESET;
+			if (p && p->flags & RZ_PRINT_FLAGS_RAINBOW) {
+				// pre = rz_cons_rgb_str_off (rgbstr, addr);
+				if (p->cons && p->cons->rgbstr) {
+					static char rgbstr[32];
+					pre = p->cons->rgbstr(rgbstr, sizeof(rgbstr), addr);
+				}
+			}
+			if (dec) {
+				rz_strbuf_appendf(sb, "%s%s%" PFMT64d "%s%c", pre, white, addr, fin, ch);
+			} else {
+				if (p && p->wide_offsets) {
+					// TODO: make %016 depend on asm.bits
+					rz_strbuf_appendf(sb, "%s0x%016" PFMT64x "%s%c", pre, addr, fin, ch);
+				} else {
+					rz_strbuf_appendf(sb, "%s0x%08" PFMT64x "%s%c", pre, addr, fin, ch);
+				}
+			}
+		} else {
+			if (dec) {
+				rz_strbuf_appendf(sb, "%s%" PFMT64d "%c", white, addr, ch);
+			} else {
+				if (p && p->wide_offsets) {
+					// TODO: make %016 depend on asm.bits
+					rz_strbuf_appendf(sb, "0x%016" PFMT64x "%c", addr, ch);
+				} else {
+					rz_strbuf_appendf(sb, "0x%08" PFMT64x "%c", addr, ch);
+				}
+			}
+		}
+	}
+}
+
+static inline void print_section(RzStrBuf *sb, RzPrint *p, ut64 at) {
```

Comment:
I think you can reuse this here in this PR to rewrite `rz_print_section` in terms of this.

Reply:
Something like
```c
RZ_API void rz_print_section(RzPrint *p, ut64 at) {
   RzStrBuf sb;
   rz_strbuf_init(&sb);
   print_section(&sb, p, at);
   char *s = rz_strbuf_drain(&sb);
   p->cb_printf("%s", s);
   free(s);
   return;
}
```